docs/architecture/deployment-execution/steps.md

```diff
@@ -0,0 +1,254 @@
+# Deployment Steps
+
+Deployments in Pulumi happens as a series of `Step`s. These are generated by the
+[step generator](https://pulumi-developer-docs.readthedocs.io/latest/docs/architecture/deployment-execution/resource-registration.html#step-generation),
+and then executed by the deployment executor.
+
+When a [Step](https://pkg.go.dev/github.com/pulumi/pulumi/pkg/v3/resource/deploy#Step) is
+applied using the `Apply()` function, it interacts with backing providers to produce the
+function that will ultimately register the resource against Pulumi's desired state graph.
+
+Each step also has an old and a new state, the old state representing the state
```

Comment:
Steps can run after each other so the old state might not be from the state file, it might be the result of the last step.